### PR TITLE
trace: add SRC_ENABLE_NET_TRACE to disable net/trace

### PR DESCRIPTION
**Pull request by bobheadxi**
This change adds `SRC_ENABLE_NET_TRACE`, which if set to `false` will disable `net/trace` support in `internal/trace` and our debug routers. The plan is to just set this to `false` in Sourcegraph.com (https://github.com/sourcegraph/deploy-sourcegraph-cloud/pull/17937) until we can absolutely confirm that `net/trace` is an issue - it's also possible this is only problematic for Sourcegraph.com levels of traffic. Follow-ups will be in [INC-214](https://sourcegraph.slack.com/archives/C05CGC5UL3E).

Context: today we noticed unreliability in Sourcegraph.com, [INC-214](https://sourcegraph.slack.com/archives/C05CGC5UL3E), where a lot of requests would get no responses. We diagnosed the issue to be a frontend pod that had a huge goroutine leak, causing unreliability in sourcegraph.com for requests routed to the affected pod. Removing the affected pod immediately resolved the stability issues.

image
image

It appears this may be caused by increased usage of `internal/trace`, which [creates a trace with `golang.org/x/net/trace`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@f3b0276dfe24018db5f76d5d8d430db98e59f2dd/-/blob/internal/trace/tracer.go?L32-33), which then [acquires a global mutex](https://sourcegraph.com/go/golang.org/x/net/-/blob/trace/trace.go?L369-380). Cloud Profile indicates this trend started ~June 5, which _somewhat_ correlates with a series of changes that removed OpenTracing usages and replaced them with OpenTelemetry/`internal/trace`, the latter of which ends up in this code path.

image
image

See incident discussion thread for more details: https://sourcegraph.slack.com/archives/C05CGC5UL3E/p1686712614273539?thread_ts=1686710733.750869&cid=C05CGC5UL3E

## Test plan

`sg start`, code review, CI

### CODE REVIEW

File: internal/trace/tracer.go [hunk 1]
```diff
@@ -7,8 +7,14 @@ import (
 	"go.opentelemetry.io/otel/attribute"
 	oteltrace "go.opentelemetry.io/otel/trace"
 	nettrace "golang.org/x/net/trace"
+
+	"github.com/sourcegraph/sourcegraph/internal/env"
 )
 
+// EnableNetTrace toggles golang.org/x/net/trace support (exported via
+// '/debug/requests' and '/debug/events') in internal/trace through SRC_ENABLE_NET_TRACE.
+var EnableNetTrace = env.MustGetBool("SRC_ENABLE_NET_TRACE", true, "Enable golang.org/x/net/trace")
```
Comment (eseliger) [inline internal/trace/tracer.go:16]: should this be off by default?

**Review by eseliger — APPROVED** [review]

File: internal/trace/tracer.go (hunk 1 above)
Comment (bobheadxi) [reply internal/trace/tracer.go:16]: I updated the PR description: 

> The plan is to just set this to false in Sourcegraph.com until we can absolutely confirm that net/trace is an issue.

@camdencheek will be taking a closer look tomorrow: https://sourcegraph.slack.com/archives/C05CGC5UL3E/p1686715265529789?thread_ts=1686710733.750869&cid=C05CGC5UL3E

**Comment by keegancsmith** [comment]
that explains why the net/trace page has become much more noisy. Personally I do rely on net/trace, I was just using it this morning. I wonder if me using net/trace was the issue, since the lock is likely held to render the trace pages.

If net/trace can't handle the traffic, I don't have much hope for other tracers. I suspect we have some requests which generates thousands of spans regularly then. This is likely the canary in the coal mine that something needs to be adjusted.

**Comment by bobheadxi** [comment]
> If net/trace can't handle the traffic, I don't have much hope for other tracers. I suspect we have some requests which generates thousands of spans regularly then. 

Camden has some hypotheses here: https://sourcegraph.slack.com/archives/C05CGC5UL3E/p1686751555470649 - it seems possible that a recovered panic left us in a bad state with net/trace, which explains the pattern of the issue we saw yesterday.

Note that net/trace is currently only disabled for frontend (see https://github.com/sourcegraph/deploy-sourcegraph-cloud/pull/17937, so other services still have it available), and the flag can be easily toggled later. I agree that this might be a poor workaround and/or not even the real issue, and that excessive tracing has become an issue as well, so let's continue discussing this in the incident channel :)

**Comment by camdencheek** [comment]
> since the lock is likely held to render the trace pages.

That seems like a likely situation. I hadn't considered that. It looks like rendering the debug page does, in fact, hold the lock. And if we have a huge number of traces, that could take a while.

> If net/trace can't handle the traffic, I don't have much hope for other tracers

FWIW, most other tracers only need to render the spans for a single request, and it handles those all independently. `x/net/trace` pulls info from all active requests, so there is a lot more "global" stuff going on, and it happens in-process so it interferes with the rest of the things the service is doing. 

@bobheadxi, do we have any insight into the volume and size of spans being exported? We need to make sure we're being smart about what is actually getting traced.

**Comment by camdencheek** [comment]
@keegancsmith, how do you use `x/net/trace`? I think we could significantly reduce its load if we just captured timings rather than events. Or even just captured event names rather than the full lazy-formatted content.

As a separate thing, we should do an audit of what we're adding as events. I think sometimes we are logging things that are extremely large (e.g. multierror composed of 100s of other errors).

**Comment by bobheadxi** [comment]
> do we have any insight into the volume and size of spans being exported? We need to make sure we're being smart about what is actually getting traced.

Well we get throttled by Honeycomb on what seems like a daily basis, and in GCP we have a _lot_:
 
image

(not sure what's up with units, but [we are charged per span](https://cloud.google.com/trace#pricing))

**Comment by keegancsmith** [comment]
> @keegancsmith, how do you use `x/net/trace`? I think we could significantly reduce its load if we just captured timings rather than events. Or even just captured event names rather than the full lazy-formatted content.

I always use the events to dive into why a specific search errored or took 10s>.

I think the issue here is we used it in a context where it had too much data and too high an event rate. It is perfectly acceptable to use where there is one trace per search / other important user initiated request.

I've used it on customer calls where they didn't have tracing enabled. It was exceptionally useful for diagnosing quickly what search is causing issues / which gitserver commands are slow / etc.

I'd be happy to remove this automatic integration and rather opt-in to tracing to net/trace / manually using net/trace like we do for honeycomb (mostly).

> As a separate thing, we should do an audit of what we're adding as events. I think sometimes we are logging things that are extremely large (e.g. multierror composed of 100s of other errors). [An example](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@436141cbe124a48f2e02b45c10009ead657b7d02/-/blob/cmd/repo-updater/repoupdater/server.go?L308-308).

Yes! In general I think our over use of multierror is an anti-pattern. Devs should return useful errors. Additionally I am sure there are quite a few example of large trace attributions. EG generated sql queries, sql arguments, etc. Would be great if we had a way to diagnose that.

**Pull request merged**